interfaces/Portalicious/src/app/services/api.service.ts

```diff
@@ -0,0 +1,133 @@
+      headers = headers.delete('Content-Type');
+    }
+
+    return headers;
+  }
+
+  private handleError(error: HttpErrorResponse) {
+    if (
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-enum-comparison
+      error.status === HttpStatusCode.TooManyRequests &&
+      !this.isRateLimitErrorShown
+    ) {
+      this.isRateLimitErrorShown = true;
+      window.alert('Rate limit exceeded. Please try again later.');
+      this.isRateLimitErrorShown = false;
+      return of(new Error('Rate limit exceeded'));
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-enum-comparison
+    if (error.status === HttpStatusCode.Unauthorized) {
+      // XXX: eventually this should be replaced with the USER_KEY exported by the AuthService
+      const rawUser = localStorage.getItem('USER_KEY');
+
+      if (!rawUser) {
+        return of(error);
+      }
+
+      const user = JSON.parse(rawUser) as User;
+
+      if (user.expires) {
+        const expires = Date.parse(user.expires);
+
+        if (expires < Date.now()) {
+          return of(new Error('Token expired'));
+        }
+      }
+
+      return of(new Error('Not authorized'));
+    }
+    return of(error);
+  }
+
+  private async performRequest<T>(
+    method: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE',
+    url: string,
+    {
+      body,
+      responseAsBlob = false,
+      isUpload = false,
+      params,
+    }: {
+      body?: unknown;
+      responseAsBlob?: boolean;
+      isUpload?: boolean;
+      params?: HttpParams;
+    } = {},
+  ): Promise<T> {
+    console.log(`ApiService ${method}: ${url}`, body ?? '');
+
+    try {
+      const response = await lastValueFrom<T | HttpErrorResponse | Error>(
+        this.http
+          .request(method, url, {
+            headers: this.createHeaders(isUpload),
+            responseType: responseAsBlob ? 'blob' : undefined,
+            withCredentials: true,
+            params,
+            body,
+          })
+          .pipe(
+            tap((response) => {
+              console.log(
+                `ApiService ${method}: ${url}${
+                  params ? `\nParams ${params.toString()}` : ''
+                }${body ? `\nBody: ${JSON.stringify(body, null, 2)}` : ''}`,
+                '\nResponse:',
+                response,
+              );
+            }),
+            catchError((error: HttpErrorResponse) => {
+              return this.handleError(error);
+            }),
+          ),
+      );
+      if (response instanceof Error || response instanceof HttpErrorResponse) {
+        // eslint-disable-next-line @typescript-eslint/only-throw-error
+        throw response;
+      }
+
+      return response;
+    } catch (error) {
+      console.error(`Error in ApiService ${method}:`, error);
+      throw error;
+    }
+  }
+
+  getVersionInfo() {
+    return this.performRequest<VersionInfo>(
+      'GET',
+      `${environment.url_121_service_api}/health/version`,
+    );
+  }
```

Comment:
My thinking is that given that `api.service` can be radically simplified compared to what is in the Portal (by removing method-specific functions and repetition), we could have a single `ApiService` that does what the `ProgramApiService` does on the existing Portal.